src/solver.jl

```diff
@@ -9,79 +10,106 @@
+    @req actions(::M, ::S)
+    as = actions(mdp)
+    ss = states(mdp)
+    @req length(::typeof(ss))
+    @req length(::typeof(as))
+    a = first(as)
+    s = first(ss)
+    dist = transition(mdp, s, a)
+    D = typeof(dist)
+    @req support(::D)
+    @req pdf(::D,::S)
+
+    @req stage(::S)
+    E = typeof(stage(s))
+    # @req HorizonLength(::M)
+    @req stage_states(::M, ::Int64)
+    @req stage_stateindex(::M, ::S, ::E)
+    @req horizon(::M)
 end
 
+"""
+    addstagerecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy, stage)
 
-# Global variable for storing number of epoch in order to pass it to functions from outer solvers
-# Is there a better way to achieve this?
-fhepoch = -1
-
-
-# Is it possible to change name of this function to solve?
-# The problem is that I am not able to use for example DiscreteValueIteration.solve() as I do not know the Solver in advance
+Store record for given stage results to `FiniteHorizonPolicy` and return updated version.
+"""
+function addstagerecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy, stage)    
+    fhpolicy.qmat[stage, :, :] = qmat
+    fhpolicy.util[stage, :] = util
+    fhpolicy.policy[stage, :] = policy
+    return fhpolicy
+end
 
-# MDP given horizon 5 assumes that agent can move 4 times
-function solve(mdp::MDP; verbose::Bool=false, new_VI::Bool=true)
-    fhpolicy = FiniteHorizonValuePolicy(mdp)
-    util = fill(0., mdp.no_states) # XXX not all MDPs have a no_states field. Suggest using length(states(mdp))
+# MDP given horizon 5 assumes that agent can move 5 times
+function POMDPs.solve(solver::FiniteHorizonSolver, m::MDP)
+    if typeof(HorizonLength(m)) == InfiniteHorizon
+        throw(ArgumentError("Argument m should be valid Finite Horizon MDP with methods from FiniteHorizonPOMDPs.jl/src/interface.jl implemented. If you are completely sure that you implemented all of them, you should also check if you have defined HorizonLength(::Type{<:MyFHMDP})"))
+    end
 
-    for epoch=mdp.horizon-1:-1:1
-        # Store number of epoch to global variable in order to work properly
-        # Is there a better way to achieve this?
-        global fhepoch = epoch
+    fhpolicy = FiniteHorizonValuePolicy(m)
+    util = fill(0., length(stage_states(m, 1)))
 
-        if verbose
-            println("EPOCH: $epoch")
+    for stage=m.horizon:-1:1
+        if solver.verbose
+            println("Stage: $stage")
         end
 
-        stage_q, util, pol = valueiterationsolver(mdp, epoch, util)
+        stage_q, util, pol = valueiterationsolver(m, stage, util)
 
-        fhpolicy = addepochrecord(fhpolicy, stage_q, util, pol)
+        fhpolicy = addstagerecord(fhpolicy, stage_q, util, pol, stage)
 
-        if verbose
+        if solver.verbose
             println("POLICY\n")
             println("QMAT")
             println(stage_q)
             println("util")
             println(util)
             println("policy")
             println(policy)
```

Comment:
Should be `fhpolicy`. There probably should be a test where the `verbose` flag is true.

Reply:
I agree this needs an update. What is your proposal for such a test? Wouldn't it be more visual test than an automated one?

Reply:
I think just having a small test where you *run* the code and don't test for anything with `verbose=true` would be enough, just to see that the code is not throwing exceptions (which it does in this case as `policy` is undefined).

Reply:
Good point, would not notice that bug without you, thank you!